src/vmm/src/persist.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Defines state structures for saving/restoring a Firecracker microVM.
+
+use devices::virtio::{
+    block::persist::BlockState, net::persist::NetState, persist::MmioTransportState,
+    vsock::persist::VsockState, Block, MmioTransport, Net, Vsock, VsockUnixBackend, TYPE_BLOCK,
+    TYPE_NET, TYPE_VSOCK,
+};
+use snapshot::Persist;
+use versionize::{VersionMap, Versionize, VersionizeError, VersionizeResult};
+use versionize_derive::Versionize;
+
+/// Holds information related to mmio space.
+#[derive(Debug, PartialEq, Versionize)]
+pub struct MmioState {
```

Comment:
Since `irq` isn't really part of `mmio`, I would reorganize the hierarchy a bit:

```rust
pub struct VmmResourcesState {
    /// Mmio base address at which the device is registered.
    pub mmio_base: u64,
    /// Mmio addr range length.
    pub len: u32,
    /// Used Irq line(s) for the device.
    // FIXME: should we support devices with multiple IRQs?
    pub irq: u32,
}

pub struct ConnectedDeviceState {
    /// Device state.
    pub device_state: SpecificDeviceState,
    /// Mmio transport state.
    pub transport_state: MmioTransportState,
    /// VmmResources.
    pub vmm_resources: VmmResourcesState,
}
```

Reply:
I would log an issue for making sure we are prepared for adding PCI and MQ features for devices. We should definetly support a vec of interrupts here. @andreeaflorescu I know there have been discussion about Interrupt abstractions in rust-vmm, can you add some feedback here so we are aligned at least from State point of view ?